Remove url-parse usage throughout repo

ChumpChief opened the pull request:
With #19840, we'll no longer use non-standard protocols in resolved URLs.  This was the remaining reason for using the url-parse library (aka Node's url support polyfill).  This library is deprecated.

This change removes all remaining usage of that package, replacing with standard `URL()`.

Aside from moving to a supported standard and away from a deprecated dependency, this also saves bundle size (looks to be about 50kb minified, comparing before/after of one of our examples).

This change revealed another defect: IParsedUrl.version claimed to permit `null` to indicate "don't load from a snapshot", but this was actually getting converted to `undefined` in `Loader.loadContainer()` (probably sensible, since we don't really support loading from no snapshot anymore or even guarantee we keep old snapshots around).  Updated the typing to reflect this reality - technically a breaking change but not really.  Fixing this is convenient, since we then don't need to worry about the return value of `URLSearchParams.get()` being `null` when a param is not found.

[AB#822](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/822)

ChumpChief committed:
Remove url-parse usage in getDiscoveredFluidResolvedUrl

ChumpChief committed:
Simplify InsecureTinyliciousUrlResolver

ChumpChief committed:
Remove use of url-parse from local-driver

ChumpChief committed:
Remove url-parse dependency from routerlicious-driver

ChumpChief committed:
Finish removing from local-driver

ChumpChief committed:
Remove from routerlicious-urlResolver

ChumpChief committed:
Remove from container-loader

ChumpChief committed:
Clarify that version can't be null, simplify resolveCore return

ChumpChief committed:
Remove remaining usages

ChumpChief committed:
Update typetest, add changeset

ChumpChief commented on packages/drivers/tinylicious-driver/src/insecureTinyliciousUrlResolver.ts:
```diff
@@ -26,11 +26,9 @@ export const defaultTinyliciousEndpoint = "http://localhost";
 	private readonly tinyliciousEndpoint: string;
 	public constructor(port = defaultTinyliciousPort, endpoint = defaultTinyliciousEndpoint) {
 		this.tinyliciousEndpoint = `${endpoint}:${port}`;
-		this.fluidProtocolEndpoint = this.tinyliciousEndpoint.replace(/(^\w+:|^)\/\//, "https://");
```
This complication wasn't necessary anymore.

ChumpChief commented on packages/loader/container-loader/src/loader.ts:
```diff
@@ -332,18 +332,17 @@ export class Loader implements IHostLoader {
 	private async resolveCore(
 		request: IRequest,
 		pendingLocalState?: IPendingContainerState,
-	): Promise<{ container: Container; parsed: IParsedUrl }> {
```
I believe the return of the parsed URL is just leftovers from when the loader used to cache containers (using the resolved url as the cache key).

ChumpChief commented on packages/loader/container-loader/src/utils.ts:
```diff
@@ -62,15 +60,20 @@ export interface IParsedUrl {
+				id: match[1],
+				path: match[2],
+				query,
+				version: parsed.searchParams.get("version") ?? undefined,
```
URLSearchParams.get() returns `null` when the param isn't found.

ChumpChief replied:
Added comment in latest

ChumpChief committed:
Parse once

ChumpChief committed:
Tinylicious uses http for localhost, not https

ChumpChief committed:
Merge remote-tracking branch 'upstream/main' into RemoveUrlParse

ChumpChief committed:
Less fake url for fluid-runner

ChumpChief committed:
Merge remote-tracking branch 'upstream/main' into RemoveUrlParse

jason-ha commented on packages/drivers/local-driver/package.json:
```diff
@@ -134,7 +134,6 @@
 		"@fluidframework/telemetry-utils": "workspace:~",
 		"events": "^3.1.0",
 		"jsrsasign": "^11.0.0",
-		"url": "^0.11.0",
```
I love it. This was something I was dealing with by a rename to avoid any uses of node features in runtime product. Full removal is nice to see.

jason-ha replied:
IntelliSense was a little odd, but it does show properly coming from either DOM or node. So, we appear all good. Back to fully loving it.

ChumpChief committed:
Merge remote-tracking branch 'upstream/main' into RemoveUrlParse

ChumpChief committed:
Merge remote-tracking branch 'upstream/main' into RemoveUrlParse

ChumpChief commented:
Hi folks!  If you're able to review I would appreciate it :)

jason-ha commented:
> Hi folks! If you're able to review I would appreciate it :)

I am also to the point where I can reconcile this with the no node dependencies PR. If you have another way to be sure there are no node dependencies type or otherwise cool.

jason-ha reviewed (approved):
Would be good to remove "url" from "import/no-nodejs-modules" allowances in lint rules.
But if not now, we should get them cleaned up in the node clean up PR.

ChumpChief committed:
Merge remote-tracking branch 'upstream/main' into RemoveUrlParse

ChumpChief committed:
PR feedback

ChumpChief committed:
Remove webpack-fluid-loader fallback too

ChumpChief committed:
Remove url usage in base_server

ChumpChief commented:
Removed the eslint exceptions, also spotted that we had a fallback in webpack-fluid-loader's webpack config that's no longer necessary.

Also found one more use in property-query-service.  I cleaned that up, but the package appears to be non-functional currently (and has probably been non-functional for some time?) - missing "debug" dependency.  Will prob follow up to understand when/if that package ever worked and what to do about it.

jason-ha commented on packages/drivers/routerlicious-urlResolver/.eslintrc.cjs:
```diff
@@ -25,7 +25,7 @@ module.exports = {
 		],
 
 		// This library is used in the browser, so we don't want dependencies on most node libraries.
-		"import/no-nodejs-modules": ["error", { allow: ["url"] }],
+		"import/no-nodejs-modules": ["error"],
```
@jason-ha, note to self, remove this when eslint-config-fluid is updated (as it will be the same).

ChumpChief committed:
Revert "Remove url usage in base_server"

This reverts commit 50f603a95bb4017430fcec88432a230c623319ec.

ChumpChief committed:
Merge remote-tracking branch 'upstream/main' into RemoveUrlParse
